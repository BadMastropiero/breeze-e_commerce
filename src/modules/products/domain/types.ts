import { z } from 'zod';

import {
  contentsApiResSchema,
  contentSchema,
  productsApiResSchema,
  productSchema,
} from './schemas';

export type Product = z.infer<typeof productSchema>;
export type Content = z.infer<typeof contentSchema>;

export type ProductsApiRes = z.infer<typeof productsApiResSchema>;
export type ContentsApiRes = z.infer<typeof contentsApiResSchema>;
