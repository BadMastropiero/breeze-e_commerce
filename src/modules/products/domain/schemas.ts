import { z } from 'zod';

export const productSchema = z.object({
  product_id: z.number(),
  title: z.string(),
  image: z.string(),
});

export const contentSchema = z.object({
  type: z.string(),
  contents: z.string(),
  position: z.string(),
});

export const contentsApiResSchema = z.object({
  data: z.array(contentSchema),
});

export const productsApiResSchema = z.object({
  products: z.array(productSchema),
});
