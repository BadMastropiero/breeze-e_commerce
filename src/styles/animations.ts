import { css, keyframes } from 'styled-components';

export const breatheAnimation = keyframes`
  0% {
    transform: translateX(-50px); /* Start position */
    opacity: 0;
  }
  100% {
    transform: translateX(0); /* End position */
    opacity: 1;
  }
`;

export const popinAnimation = keyframes`
  0% {
    opacity: 0;
    transform: scale(0.1); /* Start position (small) */
  }
  100% {
    opacity: 1;
    transform: scale(1); /* End position (normal size) */
  }
`;

export const bblFadInOutAnimation = keyframes`
  0%, 80%, 100% {
    box-shadow: 0 2.5em 0 -1.3em
  }
  40% {
    box-shadow: 0 2.5em 0 0
  }
`;

export const fadeIn = css`
  animation-name: ${breatheAnimation};
  animation-duration: 0.1s;
  animation-iteration-count: initial;
  animation-timing-function: ${(p) => p.theme.easings.base};
`;

export const popIn = css`
  animation-name: ${popinAnimation};
  animation-duration: 0.1s;
  animation-iteration-count: initial;
  animation-timing-function: ${(p) => p.theme.easings.base};
`;
