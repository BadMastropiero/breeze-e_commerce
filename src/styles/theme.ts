export const theme = {
  colors: {
    primary: {
      base: '#f02672',
      contrast: '#fff',
    },
    secondary: {
      base: '#fff',
      contrast: '#000',
    },
    darker: {
      base: '#000',
      contrast: '#fff',
    },
    lighter: {
      base: '#fff',
      contrast: '#000',
    },
    success: {
      base: '#4caf50',
      contrast: '#fff',
    },
    error: {
      base: '#f44336',
      contrast: '#fff',
    },
  },
  layout: {
    maxWidth: '1400px',
    maxWidthS: '600px',
    vGap: '10px',
    hGap: '10px',
  },
  shadows: {
    base: '0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -2px rgba(0, 0, 0, 0.05)',
    hover: '0 10px 15px -3px rgba(0, 0, 0, 0.2), 0 4px 6px -2px rgba(0, 0, 0, 0.1)',
  },
  fontSizes: {
    xl: '30px',
    l: '25px',
    m: '20px',
    s: '16px',
    sx: '14px',
  },
  borderRadius: {
    base: '0px',
    card: '12px',
  },
  easings: {
    wiggle: 'cubic-bezier(0.74, -0.6, 0.225, 1.59)',
    base: 'cubic-bezier(0.075, 0.82, 0.165, 1)',
  },
};
