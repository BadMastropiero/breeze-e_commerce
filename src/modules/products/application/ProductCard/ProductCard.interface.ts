import { Product } from '../../domain/types';

export interface ProductCardProps {
  item: Product;
  onClick?: () => void;
}
