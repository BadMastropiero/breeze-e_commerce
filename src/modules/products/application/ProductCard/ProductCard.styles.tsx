import styled from 'styled-components';

import { fadeIn } from '../../../../styles/animations';

export const StyledCard = styled.div`
  border-radius: ${(p) => p.theme.borderRadius.card};
  overflow: hidden;
  position: relative;
  cursor: pointer;
  //height: 190px;
  width: 190px;
  height: fit-content;
  box-shadow: ${(p) => p.theme.shadows.base};
  transition: box-shadow 0.2s ${(p) => p.theme.easings.base};
  margin: auto auto 1rem;

  ${fadeIn}
  &:hover {
    box-shadow: ${(p) => p.theme.shadows.hover};
  }
`;

export const StyledCardImage = styled.img`
  width: 100%;
  object-fit: cover;
  object-position: center;
  filter: grayscale(100%);
  transition: filter 0.8s ${(p) => p.theme.easings.base};

  ${StyledCard}:hover & {
    filter: grayscale(0%);
  }
`;

export const StyledCardTitle = styled.h3`
  padding: ${(p) => p.theme.layout.vGap} ${(p) => p.theme.layout.hGap};
`;
