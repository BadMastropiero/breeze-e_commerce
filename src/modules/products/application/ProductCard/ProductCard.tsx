import { FALLBACK_IMAGE_URL } from '../../../../config';
import { ProductCardProps } from './ProductCard.interface';
import { StyledCard, StyledCardImage, StyledCardTitle } from './ProductCard.styles';

function ProductCard({ item, onClick }: ProductCardProps) {
  return (
    <>
      <StyledCard data-testid="styled-card" onClick={onClick}>
        <StyledCardImage
          src={item.image}
          loading="lazy"
          alt=""
          width={190}
          height={190}
          onError={(ev) => {
            ev.currentTarget.src = FALLBACK_IMAGE_URL;
          }}
        />
        <StyledCardTitle>{item.title}</StyledCardTitle>
      </StyledCard>
    </>
  );
}

export default ProductCard;
