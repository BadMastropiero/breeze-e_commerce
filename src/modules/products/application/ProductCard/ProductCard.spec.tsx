import '@testing-library/jest-dom';

import { fireEvent, render, screen } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { describe, expect, it } from 'vitest';

import { theme } from '../../../../styles/theme';
import { Product } from '../../domain/types';
import ProductCard from './ProductCard';

const mockProduct: Product = {
  product_id: 0,
  title: '',
  image: 'https://picsum.photos/200/300',
};

describe('ProductCard', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <ThemeProvider theme={theme}>
        <ProductCard item={mockProduct} />
      </ThemeProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('triggers on click event', () => {
    let clicked = false;

    render(
      <ThemeProvider theme={theme}>
        <ProductCard
          item={mockProduct}
          onClick={() => {
            clicked = true;
          }}
        />
      </ThemeProvider>,
    );
    const card = screen.getByTestId('styled-card');
    fireEvent.click(card);
    expect(clicked).toBe(true);
  });
});
