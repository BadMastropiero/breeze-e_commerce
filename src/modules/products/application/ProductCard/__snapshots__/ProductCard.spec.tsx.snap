// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductCard > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfCDU fhyPVj"
    data-testid="styled-card"
  >
    <img
      alt=""
      class="sc-gsTDqH hjnnKk"
      height="190"
      loading="lazy"
      src="https://picsum.photos/200/300"
      width="190"
    />
    <h3
      class="sc-dlfmHC iYLPpy"
    />
  </div>
</DocumentFragment>
`;
